File: ios/ShaarliOS/SettingsVC.m
//
// SettingsVC.m
// ShaarliOS
//
// Created by Marcus Rohrmoser on 23.07.15.
// Copyright (c) 2015-2016 Marcus Rohrmoser http://mro.name/me. All rights reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#import "SettingsVC.h"
#import "MainVC.h"
#import "OnePasswordExtension.h"
#import "NSBundle+MroSemVer.h"


@interface SettingsVC() <UITextFieldDelegate, UIWebViewDelegate>
@property (weak, nonatomic) IBOutlet UITextField *txtEndpoint;
@property (weak, nonatomic) IBOutlet UISwitch *swiSecure;
@property (weak, nonatomic) IBOutlet UITextField *txtUserName;
@property (weak, nonatomic) IBOutlet UITextField *txtPassWord;
@property (weak, nonatomic) IBOutlet UILabel *lblDefaultPrivate;
@property (weak, nonatomic) IBOutlet UISwitch *swiPrivateDefault;
@property (weak, nonatomic) IBOutlet UILabel *lblTitle;
@property (weak, nonatomic) IBOutlet UISwitch *swiTags;
@property (weak, nonatomic) IBOutlet UITextField *txtTags;
@property (strong, nonatomic) IBOutlet UIActivityIndicatorView *spiLogin;

// https://github.com/AgileBits/onepassword-app-extension#use-case-1-native-app-login
@property (weak, nonatomic) IBOutlet UIButton *btn1Password;

@property (weak, nonatomic) IBOutlet UITableViewCell *cellAbout;
@property (weak, nonatomic) IBOutlet UIWebView *wwwAbout;
@end

@implementation SettingsVC


#pragma UIViewController


-(void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    NSParameterAssert(self.lblTitle);
    NSParameterAssert(self.txtEndpoint);
    NSParameterAssert(self.swiSecure);
    NSParameterAssert(self.txtUserName);
    NSParameterAssert(self.txtPassWord);
    NSParameterAssert(self.swiPrivateDefault);
    NSParameterAssert(self.lblDefaultPrivate);
    NSParameterAssert(self.swiTags);
    NSParameterAssert(self.txtTags);
    NSParameterAssert(self.wwwAbout.scrollView);
    NSParameterAssert(self.cellAbout);
    NSParameterAssert(self.spiLogin);
    [self.tableView addSubview:self.spiLogin];

    self.wwwAbout.scrollView.scrollEnabled = NO;
    self.wwwAbout.scrollView.bounces = NO;
    [self.wwwAbout loadRequest:[NSURLRequest requestWithURL:[[NSBundle bundleForClass:[self class]] URLForResource:@"about" withExtension:@"html"]]];

    self.btn1Password.enabled = [[OnePasswordExtension sharedExtension] isAppExtensionAvailable];
    self.btn1Password.alpha = self.btn1Password.enabled ? 1.0 : 0.5;
}


-(void)viewWillAppear:(BOOL)animated
{
    MRLogD(@"", nil);
    [super viewWillAppear:animated];
    NSParameterAssert(self.shaarli);

    self.lblTitle.text = self.shaarli.title;
    self.lblTitle.textColor = self.lblTitle.text ? self.txtUserName.textColor : [UIColor redColor];
    self.lblTitle.text = self.lblTitle.text ? self.lblTitle.text : NSLocalizedString(@"Not connected yət.", @"SettingsVC");

    self.txtEndpoint.text = self.shaarli.endpointStr;
    self.swiSecure.on = self.shaarli.endpointSecure;
    self.swiSecure.enabled = NO;
    self.txtUserName.text = self.shaarli.userName;
    self.txtPassWord.text = self.shaarli.passWord;
    self.swiPrivateDefault.on = self.shaarli.privateDefault;
    self.swiTags.on = self.shaarli.tagsActive;
    self.txtTags.text = self.shaarli.tagsDefault;

    [self.spiLogin stopAnimating];

    self.title = NSLocalizedString(@"Settings", @"SettingsVC");
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(actionCancel:)];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(actionSignIn:)];
}


#pragma UITextFieldDelegate


/** called when 'return' key pressed. return NO to ignore. */
-(BOOL)textFieldShouldReturn:(UITextField *)textField
{
    MRLogD(@"%@", textField.text, nil);
    if( textField == self.txtEndpoint )
        [self.txtUserName becomeFirstResponder];
    else if( textField == self.txtUserName )
        [self.txtPassWord becomeFirstResponder];
    else if( textField == self.txtPassWord )
        [self.txtTags becomeFirstResponder];
    else if( textField == self.txtTags )
        [self actionSignIn:textField];  // dispatch async?
    return YES;
}


#pragma mark Actions


-(IBAction)actionCancel:(id)sender
{
    MRLogD(@"-", nil);
    [self.navigationController popViewControllerAnimated:YES];
}


-(IBAction)actionSignIn:(id)sender
{
    MRLogD(@"-", nil);
    const BOOL wasSetUp = self.shaarli.isSetUp;

    if( self.swiTags.on ) {
        NSMutableArray *a = [NSMutableArray arrayWithCapacity:10];
        [self.txtTags.text stringByStrippingTags:a];
        self.txtTags.text = [@"#" stringByAppendingString:[a componentsJoinedByString:@" #"]];
    } else
        self.txtTags.text = @"";

    // spinner purposely covers all screen, so all other buttons are blocked meanwhile.
    self.spiLogin.frame = self.tableView.bounds;
    [self.spiLogin startAnimating];
    [self.shaarli updateEndpoint:self.txtEndpoint.text secure:self.swiSecure.on user:self.txtUserName.text pass:self.txtPassWord.text privateDefault:self.swiPrivateDefault.on tagsActive:self.swiTags.on tagsDefault:self.txtTags.text completion:^(ShaarliM * me, NSError * error) {
         dispatch_async (dispatch_get_main_queue (), ^{
                             MRLogD (@"-", nil);
                             [self.spiLogin stopAnimating];
                             if( error ) {
                                 UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString (@"Connection failed", @"SettingsVC") message:error.localizedDescription preferredStyle:UIAlertControllerStyleAlert];
                                 [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString (@"Cancel", @"SettingsVC") style:UIAlertActionStyleCancel handler:nil]];
                                 [self presentViewController:alert animated:YES completion:nil];
                             } else if( !wasSetUp && self.shaarli.isSetUp ) {
                                 UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString (@"Setup done", @"SettingsVC") message:[NSString stringWithFormat:NSLocalizedString (@"Nice, now you can activate the extension in 'Activities' and post links to %@.", @"SettingsVC"), self.shaarli.title] preferredStyle:UIAlertControllerStyleAlert];
                                 [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString (@"OK", @"SettingsVC") style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
                                                       [self.navigationController popViewControllerAnimated:YES];
                                                   }
                                  ]];
                                 [self presentViewController:alert animated:YES completion:nil];
                                 return;
                             }
                             [self.navigationController popViewControllerAnimated:YES];
                         }
                         );
     }
    ];
}


//
-(IBAction)actionFindLoginFrom1Password:(id)sender
{
    // TODO: lookup may have to be tried 2x as the url scheme (http vs. https) isn't known yet.
    [[OnePasswordExtension sharedExtension] findLoginForURLString:self.txtEndpoint.text forViewController:self sender:sender completion:^(NSDictionary * loginDictionary, NSError * error) {
         if( loginDictionary.count == 0 ) {
             if( error.code != AppExtensionErrorCodeCancelledByUser ) {
                 NSLog (@"Error invoking 1Password App Extension for find login: %@", error);
             }
             return;
         }
         self.txtUserName.text = loginDictionary[AppExtensionUsernameKey];
         self.txtPassWord.text = loginDictionary[AppExtensionPasswordKey];
     }
    ];
}



#pragma mark UIWebViewDelegate


-(void)webViewDidFinishLoad:(UIWebView *)webView
{
    MRLogD(@"%.0f", webView.scrollView.contentSize.height, nil);
    // self.cellAbout.
    NSString *ret = [webView stringByEvaluatingJavaScriptFromString:[NSString stringWithFormat:@"injectVersion('%@');", [NSBundle semVer]]];
    NSParameterAssert([@"" isEqualToString:ret]);
    MRLogD(@"%@", ret);
}


-(BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType
{
    MRLogD(@"%@", request, nil);
    NSString *scheme = request.URL.scheme;
    if( NSNotFound != [@[@"file"] indexOfObject:scheme] )
        return YES;
    if( [[UIApplication sharedApplication] canOpenURL:request.URL] )
        [[UIApplication sharedApplication] openURL:request.URL];
    return NO;
}


@end

File: swift4/ShaarliOS/SettingsVC.swift
//
//  SettingsVC.swift
//  ShaarliOS
//
//  Created by Marcus Rohrmoser on 21.02.20.
//  Copyright © 2020-2020 Marcus Rohrmoser mobile Software http://mro.name/me. All rights reserved.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import UIKit
import WebKit

internal func endpoints(_ base : String?, _ uid : String?, _ pwd : String?) -> ArraySlice<URL> {
    var urls = ArraySlice<URL>()
    guard let base = base?.trimmingCharacters(in:.whitespacesAndNewlines)
        else { return urls }
    let base_ = base.hasPrefix(HTTP_HTTPS + "://")
    ? String(base.dropFirst(HTTP_HTTPS.count+"://".count))
    : base.hasPrefix(HTTP_HTTP + "://")
    ? String(base.dropFirst(HTTP_HTTP.count+"://".count))
    : base.hasPrefix("//")
    ? String(base.dropFirst("//".count))
    : base

    guard var ep = URLComponents(string:"//\(base_)")
        else { return urls }
    ep.user = uid
    ep.password = pwd

    ep.scheme = HTTP_HTTPS; urls.append(ep.url!)
    ep.scheme = HTTP_HTTP;  urls.append(ep.url!)

    let php = "/index.php"
    let pa = ep.path.dropLast(ep.path.hasSuffix(php)
        ? php.count
        : ep.path.hasSuffix("/")
        ? 1
        : 0)

    ep.path = pa + php
    ep.scheme = HTTP_HTTPS; urls.append(ep.url!)
    ep.scheme = HTTP_HTTP;  urls.append(ep.url!)

    return urls
}


class SettingsVC: UITableViewController, UITextFieldDelegate, WKNavigationDelegate {
    @IBOutlet private var txtEndpoint       : UITextField!
    @IBOutlet private var sldTimeout        : UISlider!
    @IBOutlet private var txtTimeout        : UITextField!
    @IBOutlet private var swiSecure         : UISwitch!
    @IBOutlet private var txtUserName       : UITextField!
    @IBOutlet private var txtPassWord       : UITextField!
    @IBOutlet private var lblDefaultPrivate : UILabel!
    @IBOutlet private var swiPrivateDefault : UISwitch!
    @IBOutlet private var lblTitle          : UILabel!
    @IBOutlet private var txtTags           : UITextField!
    @IBOutlet private var spiLogin          : UIActivityIndicatorView!
    @IBOutlet private var cellAbout         : UITableViewCell!

    private let wwwAbout                    = WKWebView()
    var current                             : BlogM?

    // MARK: - Lifecycle

    // https://www.objc.io/blog/2018/04/24/bindings-with-kvo-and-keypaths/
    override func viewDidLoad() {
        super.viewDidLoad()

        title = NSLocalizedString("Settings", comment:"SettingsVC")
        navigationItem.leftBarButtonItem = UIBarButtonItem.init(barButtonSystemItem: UIBarButtonItem.SystemItem.cancel, target:self, action:#selector(SettingsVC.actionCancel(_:)))
        navigationItem.rightBarButtonItem = UIBarButtonItem.init(barButtonSystemItem: UIBarButtonItem.SystemItem.done, target:self, action:#selector(SettingsVC.actionSignIn(_:)))

        view.addSubview(spiLogin)
        spiLogin.backgroundColor = .clear
        txtTimeout.placeholder = NSLocalizedString("sec", comment:"SettingsVC")
        sldTimeout.minimumValue = Float(timeoutMinimumValue)
        sldTimeout.maximumValue = Float(timeoutMaximumValue)
        sldTimeout.value = Float(timeoutDefaultValue)

        guard let url = Bundle(for:type(of:self)).url(forResource:"about", withExtension:"html") else { return }
        cellAbout.contentView.addSubview(wwwAbout)
        wwwAbout.navigationDelegate = self
        wwwAbout.frame = cellAbout!.contentView.bounds.insetBy(dx: 8, dy: 8)
        wwwAbout.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        wwwAbout.contentScaleFactor = 1.0
        wwwAbout.scrollView.isScrollEnabled = false
        wwwAbout.scrollView.bounces = false
        wwwAbout.isOpaque = false // avoid white flash https://stackoverflow.com/a/15670274
        wwwAbout.backgroundColor = .black
        wwwAbout.customUserAgent = SHAARLI_COMPANION_APP_URL
        wwwAbout.loadFileURL(url, allowingReadAccessTo: url.deletingLastPathComponent())
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        spiLogin.stopAnimating()
        navigationItem.rightBarButtonItem?.isEnabled = true
        txtEndpoint.becomeFirstResponder()
        togui(current)
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        spiLogin.translatesAutoresizingMaskIntoConstraints = false
        let horizontalConstraint = spiLogin.centerXAnchor.constraint(equalTo: lblTitle.centerXAnchor)
        let verticalConstraint = spiLogin.centerYAnchor.constraint(equalTo: lblTitle.centerYAnchor)
        NSLayoutConstraint.activate([horizontalConstraint, verticalConstraint])

        guard let url = UIPasteboard.general.url else {
            return
        }
        let alert = UIAlertController(
            title:NSLocalizedString("Use URL form Clipboard", comment: "SettingsVC"),
            message:String(format:NSLocalizedString("do you want to use the URL\n'%@'\nas shaarli endpoint?", comment:"SettingsVC"), url.description),
            preferredStyle:.alert
        )
        alert.addAction(UIAlertAction(
            title:NSLocalizedString("Cancel", comment:"SettingsVC"),
            style:.cancel,
            handler:nil))
        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Yes", comment:"SettingsVC"),
            style:.default,
            handler:{(_) in self.togui(url)}))
        present(alert, animated:true, completion:nil)
    }

    private func togui(_ ur : URL?) {
        txtUserName.text        = nil
        txtPassWord.text        = nil
        txtEndpoint.text        = nil
        swiSecure.isOn          = false
        guard let ur = ur else { return }
        guard var uc = URLComponents(url:ur, resolvingAgainstBaseURL:true) else { return }
        txtUserName.text        = uc.user
        txtPassWord.text        = uc.password
        swiSecure.isOn          = HTTP_HTTPS == uc.scheme
        uc.password             = nil
        uc.user                 = nil
        uc.scheme               = nil
        guard let su = uc.url?.absoluteString.suffix(from: .init(encodedOffset:2)) else { return }
        txtEndpoint.text        = String(su)
    }

    private func togui(_ b : BlogM?) {
        guard let b = b else {
            lblTitle.text = NSLocalizedString("No Shaarli yet.", comment:"SettingsVC")
            lblTitle.textColor = .red
            return
        }

        lblTitle.text = b.title;
        lblTitle.textColor = txtUserName.textColor

        togui(b.endpoint)

        sldTimeout.value        = Float(b.timeout)
        sldTimeoutChanged(self)
        swiPrivateDefault.isOn  = b.privateDefault
        txtTags.text            = b.tagsDefault
    }

    // MARK: - Navigation

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let vc = segue.destination as? MainVC else {return}
        vc.current = current
    }

    // MARK: - Actions

    @IBAction func sldTimeoutChanged(_ sender: Any) {
        txtTimeout.text = String(format:"%.1f", sldTimeout.value)
    }

    @IBAction func actionCancel(_ sender: Any) {
        print("actionCancel \(type(of: self))")
        guard let navigationController = navigationController else { return }
        navigationController.popViewController(animated:true)
    }

    @IBAction func actionSignIn(_ sender: Any) {
        print("actionSignIn \(type(of: self))")

        lblTitle.textColor = txtUserName.textColor
        lblTitle.text = NSLocalizedString("", comment:"SettingsVC")// … probing server …
        spiLogin.startAnimating()

        let cli = ShaarliHtmlClient(AppDelegate.shared.semver)
        let tim = TimeInterval(sldTimeout.value)

        func recurse(_ urls:ArraySlice<URL>) {
            guard let cur = urls.first else {
                self.failure("Oops, something went utterly wrong.")
                return
            }
            cli.probe(cur, tim) { (ur, ti, er) in
                guard !ShaarliHtmlClient.isOk(er) else {
                    self.success(ur, ti, tim)
                    return
                }
                let res = urls.dropFirst()
                let tryNext = !res.isEmpty // todo: and not banned
                guard tryNext else {
                    self.failure(er)
                    return
                }
                recurse(res)
            }
        }

        let urls = endpoints(txtEndpoint.text, txtUserName.text, txtPassWord.text)
        spiLogin.startAnimating()
        navigationItem.rightBarButtonItem?.isEnabled = false
        recurse(urls)
    }

    // MARK: - Controller Logic

    private func success(_ ur:URL, _ ti:String, _ tim:TimeInterval) {
        let ad = ShaarliM.shared
        DispatchQueue.main.sync {
            ad.saveBlog(ad.defaults, BlogM(
                endpoint:ur,
                title:ti,
                timeout:tim,
                privateDefault:swiPrivateDefault.isOn,
                tagsDefault:txtTags.text ?? ""
            ))
            navigationController?.popViewController(animated:true)
        }
    }

    private func failure(_ er:String) {
        DispatchQueue.main.sync {
            spiLogin.stopAnimating()
            navigationItem.rightBarButtonItem?.isEnabled = true
            self.lblTitle.textColor = .red
            self.lblTitle.text = er
        }
    }

    // MARK: - UITextFieldDelegate

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        print("textFieldShouldReturn \(type(of: self))")
        switch textField {
        case txtEndpoint,
             txtTimeout:  txtUserName.becomeFirstResponder()
        case txtUserName: txtPassWord.becomeFirstResponder()
        case txtPassWord: txtTags.becomeFirstResponder()
        case txtTags: actionSignIn(textField) // keyboard doesn't show 'Done', but just in case... dispatch async?
        default: return false
        }
        return true
    }

    func textFieldDidEndEditing(_ textField: UITextField) {
        print("textFieldShouldReturn \(type(of: self))")
        switch textField {
        case txtTimeout:
                        guard let va = Float(txtTimeout.text ?? "") else {
                            sldTimeout.value = Float(timeoutDefaultValue)
                            sldTimeoutChanged(self)
                            return
                        }
                        sldTimeout.value = Float(timeoutFromDouble(Double(va)))
                        sldTimeoutChanged(self)
        default: break
        }
    }

    // MARK: - WKWebViewDelegate

    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        guard let url = navigationAction.request.url else {return}
        if "file" == url.scheme {
            decisionHandler(.allow)
            return
        }

        let app = UIApplication.shared
        if #available(iOS 10.0, *) {
            app.open(url)
        } else {
            app.openURL(url)
        }
        decisionHandler(.cancel)
    }

    func webView(_ sender:WKWebView, didFinish:WKNavigation!) {
        // even this late gives a flash sometimes: view.isOpaque = true
        let semv = AppDelegate.shared.semver
        let js = "injectVersion('\(semv)');"
        wwwAbout.evaluateJavaScript(js) { res,err in print(err as Any) }
        let s = wwwAbout.scrollView.contentSize
        cellAbout.contentView.bounds = CGRect(origin: .zero, size: s)
    }
}
